src/cpp/rtps/reader/WriterProxy.cpp

```diff
@@ -198,281 +237,347 @@ bool WriterProxy::maybe_add_changes_from_writer_up_to(const SequenceNumber_t& se
+                ch->getStatus() == ChangeFromWriterStatus_t::MISSING)
+            {
                 ++returnedValue;
+            }
         }
     }
 
     return returnedValue;
 }
 
-size_t WriterProxy::numberOfChangeFromWriter() const
+size_t WriterProxy::number_of_changes_from_writer() const
 {
-    std::lock_guard<std::recursive_mutex> guard(*mp_mutex);
-    return m_changesFromW.size();
+    std::lock_guard<std::recursive_mutex> guard(mutex_);
+    return changes_from_writer_.size();
 }
 
-SequenceNumber_t WriterProxy::nextCacheChangeToBeNotified()
+SequenceNumber_t WriterProxy::next_cache_change_to_be_notified()
 {
-    std::lock_guard<std::recursive_mutex> guard(*mp_mutex);
+    std::lock_guard<std::recursive_mutex> guard(mutex_);
 
-    if(lastNotified_ < changesFromWLowMark_)
+    if(last_notified_ < changes_from_writer_low_mark_)
     {
-        ++lastNotified_;
-        return lastNotified_;
+        ++last_notified_;
+        return last_notified_;
     }
 
     return SequenceNumber_t::unknown();
 }
+
+void WriterProxy::perform_initial_ack_nack(RTPSMessageGroup_t& buffer) const
+{
+    // Send initial NACK.
+    SequenceNumberSet_t sns(SequenceNumber_t(0, 0));
+    reader_->send_acknack(sns, buffer, *this, false);
+}
+
+void WriterProxy::perform_heartbeat_response(RTPSMessageGroup_t& buffer) const
+{
+    reader_->send_acknack(this, buffer, *this, heartbeat_final_flag_);
+}
+
+bool WriterProxy::process_heartbeat(
+        uint32_t count,
+        const SequenceNumber_t& first_seq,
+        const SequenceNumber_t& last_seq,
+        bool final_flag,
+        bool liveliness_flag,
+        bool disable_positive)
+{
+    std::unique_lock<std::recursive_mutex> wpLock(mutex_);
+
+    if (last_heartbeat_count_ < count)
+    {
+        // If it is the first heartbeat message, we can try to cancel initial ack.
+        // TODO: This timer cancelling should be checked if needed with the liveliness implementation.
+        // To keep PARTICIPANT_DROPPED event we should add an explicit participant_liveliness QoS.
+        // This is now commented to avoid issues #457 and #155
+        // initial_acknack_->cancel_timer();
+
+        last_heartbeat_count_ = count;
+        lost_changes_update(first_seq);
+        missing_changes_update(last_seq);
+        heartbeat_final_flag_ = final_flag;
+
+        //Analyze wheter a acknack message is needed:
+        if (!final_flag)
+        {
+            if (!disable_positive || are_there_missing_changes())
+            {
+                heartbeat_response_->restart_timer();
+            }
+        }
+        else if (final_flag && !liveliness_flag)
+        {
+            if (are_there_missing_changes())
+            {
+                heartbeat_response_->restart_timer();
+            }
+        }
+
+        //FIXME: livelinessFlag
+        if (liveliness_flag)//TODOG && attributes_->m_qos.m_liveliness.kind == MANUAL_BY_TOPIC_LIVELINESS_QOS)
```

Comment:
FIX and TODO still pending?

Reply:
Yes. Pending for manual liveliness implementation